Configuration cache issue with :pluginUnderTestMetadata when java-gradle-plugin has project dependency

### Expected Behavior
My build, which uses java-gradle-plugin, shouldn't fail with a deserialization error when configuration cache is enabled.

### Current Behavior
The build fails for a simple Java Gradle plugin which depends on another project when using the configuration cache.

### Context
I tried to get my own Java Gradle plugin to build with the configuration cache enabled.

Note that this issue looks similar to #16183. But while that particular issue got fixed before 7.0, what I am seeing now is reproduceable with Gradle versions 7.1.1 and 7.2-rc3.

### Steps to Reproduce
Here is minimalistic project that reproduces the issue: [my-java-gradle-plugin-cc.zip](https://github.com/gradle/gradle/files/6999298/my-java-gradle-plugin-cc.zip)

Steps to reproduce:
1. ./gradlew build
2. ./gradlew build

Results in:

> $ ./gradlew build
> Configuration cache is an incubating feature.
> Calculating task graph as no configuration cache is available for tasks: build
> 
> BUILD SUCCESSFUL in 14s
> 8 actionable tasks: 8 executed
> Configuration cache entry stored.
> 
> $ ./gradlew build
> Configuration cache is an incubating feature.
> Reusing configuration cache.
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Could not load the value of field `collection` of `org.gradle.api.internal.file.AbstractFileCollection$ElementsProvider` bean found in field `provider` of `org.gradle.configurationcache.serialization.codecs.ProviderBackedFileCollectionSpec` bean found in field `pluginClasspath` of task `:my-plugin:pluginUnderTestMetadata` of type `org.gradle.plugin.devel.tasks.PluginUnderTestMetadata`.
> > org.gradle.plugin.devel.plugins.JavaGradlePluginPlugin$$Lambda$2590/0x0000000841e73440
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> CONFIGURE FAILED in 916ms
> Configuration cache entry reused.

### Your Environment
Build scan URL:
https://gradle.com/s/xwuvoafg2ufky

--- a/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIntegrationTest.groovy
+++ b/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheIntegrationTest.groovy
@@ -16,11 +16,58 @@
 
 package org.gradle.configurationcache
 
+import org.gradle.api.Plugin
+import org.gradle.api.Project
 import org.gradle.initialization.LoadProjectsBuildOperationType
 import org.gradle.initialization.StartParameterBuildOptions.ConfigurationCacheRecreateOption
 import org.gradle.integtests.fixtures.BuildOperationsFixture
+import spock.lang.Issue
 
 class ConfigurationCacheIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    @Issue("https://github.com/gradle/gradle/issues/18064")
+    def "can build plugin with project dependencies"() {
+        given:
+        settingsFile << """
+            include 'my-lib'
+            include 'my-plugin'
+        """
+        file('my-lib/build.gradle') << """
+            plugins { id 'java' }
+        """
+        file('my-plugin/build.gradle') << """
+            plugins { id 'java-gradle-plugin' }
+
+            dependencies {
+              implementation project(":my-lib")
+            }
+
+            gradlePlugin {
+              plugins {
+                myPlugin {
+                  id = 'com.example.my-plugin'
+                  implementationClass = 'com.example.MyPlugin'
+                }
+              }
+            }
+        """
+        file('src/main/java/com/example/MyPlugin.java') << """
+            package com.example;
+            public class MyPlugin implements $Plugin.name<$Project.name> {
+              @Override
+              public void apply($Project.name project) {
+              }
+            }
+        """
+        def configurationCache = newConfigurationCacheFixture()
+
+        when:
+        configurationCacheRun "build"
+        configurationCacheRun "build"
+
+        then:
+        configurationCache.assertStateLoaded()
+    }
 
     def "can copy zipTree"() {
         given:

--- a/subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java
+++ b/subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/JavaGradlePluginPlugin.java
@@ -24,6 +24,7 @@ import org.gradle.api.Project;
 import org.gradle.api.Task;
 import org.gradle.api.artifacts.ArtifactView;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
 import org.gradle.api.artifacts.dsl.DependencyHandler;
 import org.gradle.api.file.CopySpec;
 import org.gradle.api.file.FileCollection;
@@ -68,6 +69,8 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.Callable;
+
+import static org.gradle.api.internal.lambdas.SerializableLambdas.spec;
 
 /**
  * A plugin for building java gradle plugins. Automatically generates plugin descriptors. Emits warnings for common error conditions. <p> Provides a direct integration with TestKit by declaring the
@@ -190,13 +193,7 @@ public class JavaGradlePluginPlugin implements Plugin<Project> {
                 SourceSet sourceSet = extension.getPluginSourceSet();
                 Configuration runtimeClasspath = project.getConfigurations().getByName(sourceSet.getRuntimeClasspathConfigurationName());
                 ArtifactView view = runtimeClasspath.getIncoming().artifactView(config -> {
-                    config.componentFilter(componentId -> {
-                        if (componentId instanceof OpaqueComponentIdentifier) {
-                            DependencyFactory.ClassPathNotation classPathNotation = ((OpaqueComponentIdentifier) componentId).getClassPathNotation();
-                            return classPathNotation != DependencyFactory.ClassPathNotation.GRADLE_API && classPathNotation != DependencyFactory.ClassPathNotation.LOCAL_GROOVY;
-                        }
-                        return true;
-                    });
+                    config.componentFilter(spec(JavaGradlePluginPlugin::excludeGradleApi));
                 });
                 return pluginUnderTestMetadataTask.getProject().getObjects().fileCollection().from(
                     sourceSet.getOutput(),
@@ -204,6 +201,14 @@ public class JavaGradlePluginPlugin implements Plugin<Project> {
                 );
             });
         });
+    }
+
+    private static boolean excludeGradleApi(ComponentIdentifier componentId) {
+        if (componentId instanceof OpaqueComponentIdentifier) {
+            DependencyFactory.ClassPathNotation classPathNotation = ((OpaqueComponentIdentifier) componentId).getClassPathNotation();
+            return classPathNotation != DependencyFactory.ClassPathNotation.GRADLE_API && classPathNotation != DependencyFactory.ClassPathNotation.LOCAL_GROOVY;
+        }
+        return true;
     }
 
     private void establishTestKitAndPluginClasspathDependencies(Project project, GradlePluginDevelopmentExtension extension, TaskProvider<PluginUnderTestMetadata> pluginClasspathTask) {

